Sources/SmokeHTTP1/SmokeHTTP1Server.swift

```diff
@@ -18,55 +18,223 @@
 import Foundation
 import NIO
 import NIOHTTP1
+import NIOExtras
+import LoggerAPI
 
 public struct ServerDefaults {
     static let defaultHost = "0.0.0.0"
     public static let defaultPort = 8080
-    public static let defaultServerThreads = 6
+}
+
+public enum SmokeHTTP1ServerError: Error {
+    case shutdownAttemptOnUnstartedServer
 }
 
 /**
  A basic non-blocking HTTP server that handles a request with an
  optional body and returns a response with an optional body.
  */
 public class SmokeHTTP1Server {
+    struct LifecycleState {
+        enum State {
+            case initialized
+            case running
+            case shuttingDown
+            case shutDown
+        }
+        private var shutdownCompletionHandlers: [() -> Void] = []
+        private var serverState: State = .initialized
+        private var stateLock: NSLock = NSLock()
```

Comment:
you can't put a lock into a struct to protect mutations of the struct itself. This code is already incorrect, no matter what the body of the function is. Run for example this program:

```swift
import Dispatch

struct Foo {
    mutating func mutateMe() {
    }
}

var foo = Foo()
foo.mutateMe()
let g = DispatchGroup()
for _ in 0..<10000 {
    DispatchQueue.global().async(group: g) {
        foo.mutateMe()
    }
}
g.wait()
```

with TSan:

    swiftc -O -sanitize=thread test.swift && ./test

and it'll tell you that there's a problem:

```
==================
WARNING: ThreadSanitizer: Swift access race (pid=18356)
  Modifying access of Swift variable at 0x00010027854e by thread T2:
    #0 closure #1 in  <null>:1063760 (test:x86_64+0x100001508)
    #1 thunk for @escaping @callee_guaranteed () -> () <null>:1063760 (test:x86_64+0x10000154d)
    #2 __wrap_dispatch_group_async_block_invoke <null>:1063760 (libclang_rt.tsan_osx_dynamic.dylib:x86_64h+0x6b147)
    #3 _dispatch_client_callout <null>:1063760 (libdispatch.dylib:x86_64+0x363c)

  Previous modifying access of Swift variable at 0x00010027854e by thread T1:
    #0 closure #1 in  <null>:1063760 (test:x86_64+0x100001508)
    #1 thunk for @escaping @callee_guaranteed () -> () <null>:1063760 (test:x86_64+0x10000154d)
    #2 __wrap_dispatch_group_async_block_invoke <null>:1063760 (libclang_rt.tsan_osx_dynamic.dylib:x86_64h+0x6b147)
    #3 _dispatch_client_callout <null>:1063760 (libdispatch.dylib:x86_64+0x363c)

  Location is global 'thunk for @escaping @callee_guaranteed () -> ()' at 0x000100278510 (test+0x00010000154e)

  Thread T2 (tid=535804, running) is a GCD worker thread

  Thread T1 (tid=535803, running) is a GCD worker thread

SUMMARY: ThreadSanitizer: Swift access race (test:x86_64+0x100001508) in closure #1 in 
==================
```

A struct always acts as a value and the compiler is totally free to copy it around. So if you have a function like

```swift
struct Foo {
    let lock: NSLock
    var x: Int

    mutating func mutateMe() {
        self.lock.lock()
        self.x += 1
        self.lock.unlock()
    }

    func readMe() -> Int {
        self.lock.lock()
        defer {
            self.lock.unlock()
        }
        return self.x += 1
    }
}
```

and you run

```swift
var foo = Foo()
foo.mutateMe()
```

then the compiler is totally free to do something akin to

```swift
var foo = Foo()

var foo2 = foo // copy the struct
foo2.lock.lock()
foo2.x += 1
foo2.lock.unlock()
foo = foo2 // copy it back
```

The problem now is that you still have unprotected access to the structures and you'll run into trouble.

If you want to synchronise a value, you'll need to put the lock outside of the value, like next to it.

Or you make your state a `class`, then the whole thing will work.

Reply:
Thanks! I should have noticed this 😬